rust/capture/src/passes/invalid_surrogates.rs

```diff
@@ -0,0 +1,226 @@
+    pub fn needed(input: &str) -> bool {
+        input.contains("\\u")
+    }
+
+    pub fn new(input: Chars<'a>) -> Self {
+        Self {
+            input,
+            last_seen: LastSeen::Char,
+            pending_output: Vec::with_capacity(32),
+            pending_ptr: 0,
+            escape_seq_buf: String::with_capacity(32),
+        }
+    }
+
+    fn queue(&mut self, c: char) {
+        if self.last_seen == LastSeen::Escape {
+            // When we enter an escape sequence, we swallow the backslash,
+            // to avoid having to backtrack if we drop an invalid escape sequence.
+            // So we have to emit it here.
+            self.pending_output.push('\\');
+            self.pending_output.push(c);
+            self.last_seen = LastSeen::Char;
+        } else if c == '\\' {
+            // If we're not already in an escape sequence, enter one, dropping the char to
+            // avoid needing to backtrack
+            self.last_seen = LastSeen::Escape;
+        } else {
+            // If we're not in an escape sequence, and not entering one, just push
+            self.last_seen = LastSeen::Char;
+            self.pending_output.push(c);
+        }
+    }
+
+    fn queue_str(&mut self, s: &str) {
+        for c in s.chars() {
+            self.queue(c);
+        }
+    }
+
+    fn pop(&mut self) -> Option<char> {
+        // We push chars into the buffer reading left-to-right, and need to emit them
+        // in the same order, so we have to track our stack index, and reset it when we
+        // run out of chars to pop.
+        if self.pending_ptr < self.pending_output.len() {
+            let c = self.pending_output[self.pending_ptr];
+            self.pending_ptr += 1;
+            Some(c)
+        } else {
+            self.pending_output.clear();
+            self.pending_ptr = 0;
+            None
+        }
+    }
+
+    fn step(&mut self) -> Option<char> {
+        if let Some(c) = self.pop() {
+            return Some(c);
+        }
+
+        // We're out of input, and we've go no pending output, so we're done
+        let Some(c) = self.input.next() else {
+            // If we're all out of input and the last thing we saw was an escape,
+            // we have to emit that escape character. We just do that directly here,
+            // knowing the next call around we'll return None.
+            // Note that since we're parsing strings to get turned into json values,
+            // we technically know this will be immediately discarded, but there's
+            // no harm making it "correct" first.
+            if self.last_seen == LastSeen::Escape {
+                self.last_seen = LastSeen::Char;
+                return Some('\\');
+            };
+            return None;
+        };
+
+        match (self.last_seen, c) {
+            (LastSeen::Escape, 'u') => {
+                self.escape_seq_buf.clear();
+                if collect_n_chars(&mut self.escape_seq_buf, &mut self.input, 4) < 4 {
```

Comment:
Would be nice to have a like `HEX_ESCAPE_LENGTH = 4` const if that's what the 4 is represending?

Reply:
Done